**Михаил, привет!**

Меня зовут Артемьева Саша, я буду проверять твой проект. Приятно познакомиться, хоть и заочно:)
    
Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Это как тестовое задание при приеме на работу: очень много мелких ошибок могут стать причиной отказа кандидату. 

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

# Исследование надежности заемщиков.

## Шаг 1. Обзор данных

**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание данных:**

* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита;

**Цель исследования**:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**

О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных. 

Проверим данные на ошибки и оценим их влияние на исследование. Затем, на этапе предобработки вы поищете возможность исправить самые критичные ошибки данных.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Ответы на вопросы на основе данных.

In [349]:
import pandas as pd # импорт библиотеки pandas
df = pd.read_csv('/datasets/data.csv') # чтение файла с данными и сохранение в df
display(df)
df.info() # получение общей информации о данных в таблице df

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

# Шаг 2 Предобработка данных

Посмотрим сколько именно и в каких столбцах не хватает значений.

In [350]:
display(df.isna().sum()) # подсчёт пропусков

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

2174 - значений не хватает. Найдем их долю среди всей таблицы:

In [351]:
print('Доля пропущенных значений:', round(df['children'].count() / df['days_employed'].isna().sum(),1))

Доля пропущенных значений: 9.9


Почти 10%. Проверим взаимосвязь по типу занятости.

In [352]:
display(df[df['days_employed'].isna()]['income_type'].unique())
display(df[df['total_income'].isna()]['income_type'].unique())

array(['пенсионер', 'госслужащий', 'компаньон', 'сотрудник',
       'предприниматель'], dtype=object)

array(['пенсионер', 'госслужащий', 'компаньон', 'сотрудник',
       'предприниматель'], dtype=object)

Пропуски значений есть у всех типов клиентов, вероятнее всего они сделаны не при оформлении или исправлении данных, а при выгрузке в файл.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>  Можем сообщить о проблеме заказчику и запросить метаданные для глубокого анализа

## Шаг 2.1 Заполнение пропусков

Так как столбцы days_employed и total_income имеют формат - float64, а не object, заменяем пропущенные значения на 0.

In [353]:
df['days_employed'] = df['days_employed'].fillna(0) # замена пропущенных значений на 0 в столбце days_employed 
df['total_income'] = df['total_income'].fillna(0) # замена пропущенных значений на 0 в столбце total_income 
display(df.isna().sum()) # подсчёт пропусков после замены

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Пропущенные значения days_employed и total_income заменены на 0.                                 
Теперь в заменим 0 на значения медиан, по каждому типу занятости клиента.

*Зачем нужна медиана, если есть среднее? Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства.*

Создаем функцию median_function, которая формирует df по столбцу ежемесячного дохода = 0 и по типу занятости из списка. Приравниваем к значению медианты, которую получили сгруппировав ежемесячный доход по типу занятости.             

In [354]:
def median_function(columns_list):
    income_type_list =['безработный', 'в декрете','госслужащий','компаньон','пенсионер','предприниматель','сотрудник','студент']
    for i in income_type_list:
        df.loc[(df[columns_list[0]] == 0) & (df[columns_list[1]] == i), columns_list[0]] = df.groupby(columns_list[1])[columns_list[0]].median()[i]

*Столбец ежемесячный доход:*                 

In [355]:
median_function(['total_income','income_type'])
df.groupby('income_type')['total_income'].median() # вывод медианных значений ежемесячного дохода, относительно типа занятости.

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        139034.452386
компаньон          162401.351555
пенсионер          110179.955936
предприниматель    374372.358710
сотрудник          133546.457238
студент             98201.625314
Name: total_income, dtype: float64

*Столбец общий трудовой стаж в днях*               

In [356]:
median_function(['days_employed','income_type'])
df.groupby('income_type')['days_employed'].median() # вывод медианных значений стажа, относительно типа занятости.

income_type
безработный        366413.652744
в декрете           -3296.759962
госслужащий         -2385.358043
компаньон           -1311.128244
пенсионер          360515.968234
предприниматель      -390.636063
сотрудник           -1360.363902
студент              -578.751554
Name: days_employed, dtype: float64

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b>  Эти замены элегантнее можно было сделать в цикле, создав список из нужных для замены колонок

<div class="alert alert-info">
<b> Комментарий студента:</b> Сделал более-менее элегантнее))

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b>  Молодец!

Отрицательного трудового стажа неможет быть.
Медианные значения по типу безработный и пенсионер - аномальны. Так как это медианные значения, то аномально большинство значений, при этом всем они положительны. Похоже что значения данных трудового стажа у безработных и пенсионеров, завышенно рукотворно. 

*Надеемся что пенсионерам завысили из лучших побуждений*

In [357]:
df.info() # получение общей информации о данных в измененной таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Пропуски заполнили, но теперь нужно проверить таблицу на наличие аномалий.

## Шаг 2.2 Проверка данных на аномалии и исправления.

Проверим все столбцы на наличие уникальных значений.

In [358]:
print(df.columns) # правильность названий столбцов
for row in df:
    print(df[row].head(0)) # вывод названия столбца и типа данных
    print(df[row].sort_values().unique()) # вывод уникальных значений
    print()

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')
Series([], Name: children, dtype: int64)
[-1  0  1  2  3  4  5 20]

Series([], Name: days_employed, dtype: float64)
[-18388.94990057 -17615.56326563 -16593.47281726 ... 401675.09343386
 401715.81174889 401755.40047533]

Series([], Name: dob_years, dtype: int64)
[ 0 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41
 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65
 66 67 68 69 70 71 72 73 74 75]

Series([], Name: education, dtype: object)
['ВЫСШЕЕ' 'Высшее' 'НАЧАЛЬНОЕ' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Начальное'
 'Неоконченное высшее' 'СРЕДНЕЕ' 'Среднее' 'УЧЕНАЯ СТЕПЕНЬ'
 'Ученая степень' 'высшее' 'начальное' 'неоконченное высшее' 'среднее'
 'ученая степень']

Series([], Name: education_id, dtype: int64)
[0 1 2 3 4]

Series([], Name: family_status

Первое что бросается в глаза:

* `children` — количество детей, -1 и 20. Нужно уточнить.
* `days_employed` — общий трудовой стаж, есть отрицательные значения и значения "артефакты" (401675.09343386, 401715.81174889, 401755.40047533 и т.п)
* `dob_years` — возраст клиента в годах, 0 лет. Нужно уточнить.
* `education` — уровень образования клиента, нужно привести к единому регистру.
* `gender` — пол клиента: XNA. Нужно уточнить.
* `purpose` — цель получения кредита, много однотипных целей.

Начнем самого этого. 
* `children` — количество детей, -1 и 20. Нужно уточнить.

Взглянем на данные:

In [359]:
display(df[df['children'] == -1])

children  days_employed  dob_years            education  education_id  \
291          -1   -4417.703588         46              среднее             1   
705          -1    -902.084528         50              среднее             1   
742          -1   -3174.456205         57              среднее             1   
800          -1  349987.852217         54              среднее             1   
941          -1  360505.668544         57              Среднее             1   
1363         -1   -1195.264956         55              СРЕДНЕЕ             1   
1929         -1   -1461.303336         38              среднее             1   
2073         -1   -2539.761232         42              среднее             1   
3814         -1   -3045.290443         26              Среднее             1   
4201         -1    -901.101738         41              среднее             1   
4402         -1  398001.302888         64              СРЕДНЕЕ             1   
4542         -1   -1811.899756         32              среднее             1   
5272         -1   -1143.485347         46              среднее             1   
6013         -1   -1361.258696         46               высшее             0   
6385         -1  370215.476226         48              среднее             1   
7190         -1   -5928.202068         34              СРЕДНЕЕ             1   
7278         -1    -526.318451         51              среднее             1   
7615         -1   -1360.363902         35              среднее             1   
7685         -1   -3237.360455         53              среднее             1   
8229         -1   -1803.441074         54               высшее             0   
8251         -1   -3113.998449         31              среднее             1   
8561         -1    -195.479496         31              среднее             1   
9565         -1   -2896.629686         34               высшее             0   
9574         -1   -2710.419901         28               высшее             0   
10177        -1   -1743.604011         37              среднее             1   
10365        -1  345774.125957         63               ВЫСШЕЕ             0   
10918        -1  340499.039342         50              среднее             1   
11109        -1   -3438.463024         59              среднее             1   
11268        -1   -1048.782203         30               высшее             0   
11617        -1    -370.827130         27              среднее             1   
12304        -1   -8493.101252         61  неоконченное высшее             2   
13786        -1   -1311.128244         42              среднее             1   
14357        -1    -268.337037         23              среднее             1   
15147        -1   -9851.184337         44              среднее             1   
15435        -1   -3614.220232         40               высшее             0   
16129        -1    -457.861760         33              среднее             1   
16265        -1   -2802.218127         40               высшее             0   
17095        -1   -2809.693200         34              среднее             1   
17429        -1    -895.379738         37              среднее             1   
17657        -1   -4571.957475         41              среднее             1   
18219        -1   -3575.215641         33              среднее             1   
19100        -1    -617.246968         38              среднее             1   
19366        -1    -372.034749         43              СРЕДНЕЕ             1   
19417        -1  350340.760224         28              среднее             1   
20393        -1  355157.107212         69              среднее             1   
20717        -1    -661.822321         32               высшее             0   
21140        -1   -1422.668059         44              среднее             1   

               family_status  family_status_id gender  income_type  debt  \
291         гражданский брак                 1      F    сотрудник     0   
705          ж

Похоже что эта ошибка опявилась так же как и отрицательное значение трудового стажа.

In [360]:
display(df[df['children'] == 20])

children  days_employed  dob_years education  education_id  \
606          20    -880.221113         21   среднее             1   
720          20    -855.595512         44   среднее             1   
1074         20   -3310.411598         56   среднее             1   
2510         20   -2714.161249         59    высшее             0   
2941         20   -2161.591519          0   среднее             1   
...         ...            ...        ...       ...           ...   
21008        20   -1240.257910         40   среднее             1   
21325        20    -601.174883         37   среднее             1   
21390        20   -1311.128244         53   среднее             1   
21404        20    -494.788448         52   среднее             1   
21491        20    -173.954460         27   среднее             1   

         family_status  family_status_id gender income_type  debt  \
606    женат / замужем                 0      M   компаньон     0   
720    женат / замужем                 0      F   компаньон     0   
1074   женат / замужем                 0      F   сотрудник     1   
2510    вдовец / вдова                 2      F   сотрудник     0   
2941   женат / замужем                 0      F   сотрудник     0   
...                ...               ...    ...         ...   ...   
21008  женат / замужем                 0      F   сотрудник     1   
21325  женат / замужем                 0      F   компаньон     0   
21390  женат / замужем                 0      M   компаньон     0   
21404  женат / замужем                 0      M   компаньон     0   
21491  женат / замужем                 0      F   сотрудник     0   

        total_income                                purpose  
606    145334.865002                          покупка жилья  
720    112998.738649                   покупка недвижимости  
1074   229518.537004                  получение образования  
2510   264474.835577  операции с коммерческой недвижимостью  
2941   199739.941398                  на покупку автомобиля  
...              ...                                    ...  
21008  133524.010303                        свой автомобиль  
21325  102986.065978                 профильное образование  
21390  162401.351555             покупка жилой недвижимости  
21404  156629.683642        операции со своей недвижимостью  
21491  155520.899688     на покупку подержанного автомобиля  

[76 rows x 12 columns]

1-4 семей может быть по 20 детей, но 76 семей по 20 детей в России - это явно опечатка.

*Где-то вздохнула демография...*

In [361]:
print('Доля аномалий',round((((df[df['children'] == -1]['children'].count() + df[df['children'] == 20]['children'].count()) / df['children'].count())*100),2),'%')

Доля аномалий 0.57 %


Избавляемся от них

In [362]:
df = df[(df['children'] != -1) & (df['children'] != 20)]
print(df['children'].unique()) # Проверка

[1 0 3 2 4 5]


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Сейчас доля аномалий мала и ни на что не повлияет. В следующий раз старайся сразу высчитывать их долю и если она меньше 5%, смело избавляйся от нее. Возможно наше предположение о том, что 20==2, а -1==1 неверно

</div>

<div class="alert alert-info">
<b> Комментарий студента:</b> Избавился от этой доли)

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Отлично!

Продолжим:

* `days_employed` — общий трудовой стаж, есть отрицательные значения и значения "артефакты" (401675.09343386, 401715.81174889, 401755.40047533 и т.п)

Сначала исправим "-" :

In [363]:
df['days_employed'] = abs(df['days_employed']) # исправляем "-" в отрицательных значениях
print(df.groupby('income_type')['days_employed'].median())

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2385.358043
компаньон            1318.013821
пенсионер          360632.255610
предприниматель       390.636063
сотрудник            1360.363902
студент               578.751554
Name: days_employed, dtype: float64


Теперь все значения трудового стажа положительные.

Попробуем найти все аномальные значения в столбце трудового стажа. Но так как есть и нормальные значения, отфильтровать данные по максимально возможному числу трудового стажа.

In [364]:
max_dob_teor = ((df['dob_years'].max()-18)*365)*2
# максимальный возраст в таблице - 18 лет * 365 дней(работа без выходных) * 2 (надбавка в условиях крайнего севера)
print('Максимальный трудовой стаж:',max_dob_teor,'дней') # Максимальный теоретический стаж работы, без выходных.

Максимальный трудовой стаж: 41610 дней


При условиях крайнего севера(или коэффициента работы = 2), стаж 41610 дней.

Теперь отфильтруем таблицу по этой велечине:

In [365]:
display(df[df['days_employed'] > max_dob_teor]) # Фильтрация таблицы по макс.теор.стаж

children  days_employed  dob_years education  education_id  \
4             0  340266.072047         53   среднее             1   
12            0  360505.668544         65   среднее             1   
18            0  400281.136913         53   среднее             1   
24            1  338551.952911         57   среднее             1   
25            0  363548.489348         67   среднее             1   
...         ...            ...        ...       ...           ...   
21505         0  338904.866406         53   среднее             1   
21508         0  386497.714078         62   среднее             1   
21509         0  362161.054124         59    высшее             0   
21518         0  373995.710838         59   СРЕДНЕЕ             1   
21521         0  343937.404131         67   среднее             1   

               family_status  family_status_id gender income_type  debt  \
4           гражданский брак                 1      F   пенсионер     0   
12          гражданский брак                 1      M   пенсионер     0   
18            вдовец / вдова                 2      F   пенсионер     0   
24     Не женат / не замужем                 4      F   пенсионер     0   
25           женат / замужем                 0      M   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21505       гражданский брак                 1      M   пенсионер     0   
21508        женат / замужем                 0      M   пенсионер     0   
21509        женат / замужем                 0      M   пенсионер     0   
21518        женат / замужем                 0      F   пенсионер     0   
21521        женат / замужем                 0      F   пенсионер     0   

        total_income                                purpose  
4      158616.077870                        сыграть свадьбу  
12     110179.690761                        сыграть свадьбу  
18      56823.777243     на покупку подержанного автомобиля  
24     290547.235997  операции с коммерческой недвижимостью  
25      55112.757732                   покупка недвижимости  
...              ...                                    ...  
21505   75439.993167                        сыграть свадьбу  
21508   72638.590915                           недвижимость  
21509   73029.059379               операции с недвижимостью  
21518  153864.650328                   сделка с автомобилем  
21521  155999.806512                   сделка с автомобилем  

[3841 rows x 12 columns]

Теперь посмотрим уникальные значения в столбце типа занятости:

In [366]:
display(df[df['days_employed'] > max_dob_teor]['income_type'].unique())

array(['пенсионер', 'безработный'], dtype=object)

И все таки есть только пенсионеры и безработные с аномальным стажем. Посчитаем их количество.

In [367]:
print('Аномальный трудовой стаж у безработных:',df.loc[(df['days_employed'] > max_dob_teor) & (df['income_type'] == 'безработный'), 'days_employed'].count())
print('Аномальный трудовой стаж у пенсионеров:',df.loc[(df['days_employed'] >max_dob_teor) & (df['income_type'] == 'пенсионер'), 'days_employed'].count())

Аномальный трудовой стаж у безработных: 2
Аномальный трудовой стаж у пенсионеров: 3839


Поситаем общее количество безработных и пенсионеров во всем дата-фрейме.

In [368]:
print('Общее число безработных:', df[df['income_type'] == 'безработный']['income_type'].count())
print('Общее число пенсионеро:', df[df['income_type'] == 'пенсионер']['income_type'].count())

Общее число безработных: 2
Общее число пенсионеро: 3839


Абсолютно все данные аномальны. Продолжим работу с ними после преобработки всей таблицы.

Разберемся с возрастом клиентов.

* `dob_years` — возраст клиента в годах, 0 лет. Нужно уточнить.

In [369]:
display(df[df['dob_years'] == 0])
print('Погрешность:' ,round(((df[df['dob_years'] == 0]['dob_years'].count()/df['dob_years'].count())*100),3),'%')

children  days_employed  dob_years education  education_id  \
99            0  346541.618895          0   Среднее             1   
149           0    2664.273168          0   среднее             1   
270           3    1872.663186          0   среднее             1   
578           0  397856.565013          0   среднее             1   
1040          0    1158.029561          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0    1360.363902          0   среднее             1   
20462         0  338734.868540          0   среднее             1   
20577         0  331741.271455          0   среднее             1   
21179         2     108.967042          0    высшее             0   
21313         0    1268.487728          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
99           женат / замужем                 0      F   пенсионер     0   
149                в разводе                 3      F   сотрудник     0   
270          женат / замужем                 0      F   сотрудник     0   
578          женат / замужем                 0      F   пенсионер     0   
1040               в разводе                 3      F   компаньон     0   
...                      ...               ...    ...         ...   ...   
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                                 purpose  
99      71291.522491                              автомобиль  
149     70176.435951                       операции с жильем  
270    102166.458894                            ремонт жилью  
578     97620.687042  строительство собственной недвижимости  
1040   303994.134987                         свой автомобиль  
...              ...                                     ...  
19829  133546.457238                                   жилье  
20462  259193.920299                    покупка своего жилья  
20577  129788.762899                            недвижимость  
21179  240702.007382        строительство жилой недвижимости  
21313  152419.184851                       операции с жильем  

[100 rows x 12 columns]

Погрешность: 0.467 %


Погрешность почти 0,5% мы можем даже удалить эти данные, они не повлияют на общий вывод. Но все таки мы их восстановим, через медиану по типу занятости.

In [370]:
display(df.groupby('income_type')['dob_years'].median()) # Медианные значения возраста по типу занятости

income_type
безработный        38.0
в декрете          39.0
госслужащий        40.0
компаньон          39.0
пенсионер          60.0
предприниматель    42.5
сотрудник          39.0
студент            22.0
Name: dob_years, dtype: float64

Заменяем 0 на медианные значения возраста. И посмотрим уникальные значения, убедимся что 0 нет.

In [371]:
median_function(['dob_years','income_type'])
display(df['dob_years'].sort_values().unique())

array([19., 20., 21., 22., 23., 24., 25., 26., 27., 28., 29., 30., 31.,
       32., 33., 34., 35., 36., 37., 38., 39., 40., 41., 42., 43., 44.,
       45., 46., 47., 48., 49., 50., 51., 52., 53., 54., 55., 56., 57.,
       58., 59., 60., 61., 62., 63., 64., 65., 66., 67., 68., 69., 70.,
       71., 72., 73., 74., 75.])

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Для оптимизации кода здесь также можно было бы применить цикл

<div class="alert alert-info">
<b> Комментарий студента:</b> Сделал более-менее элегантнее))

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b>  Корректно

Значения возраста получили в формате float, в будущем мы его приведем в порядок.

Раберемся с полом клиентов.
* `gender` — пол клиента: XNA. Нужно уточнить.

In [372]:
df[df['gender'] == 'XNA'] # Фильтруем таблицу по гендеру XNA

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502       24.0  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Всего 1 экземпляр. Удаляем.

*С зп 204к можно не заканчивать высшее и не определяться с гендером...*

In [373]:
df = df[df['gender'] != 'XNA'].reset_index(drop=True) # удаляем строку

Проверяем на наличие уникальных значений в гендоре.

In [374]:
print(df['gender'].unique()) # проверка на наличие значений

['F' 'M']


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Отлично!

Вернемся к аномальным значениям в стаже работы и отсортируем по возрасту.

In [375]:
display(df[df['days_employed'] > max_dob_teor].sort_values('dob_years'))

children  days_employed  dob_years education  education_id  \
1234          0  334764.259831       22.0   Среднее             1   
3604          0  360505.668544       24.0   среднее             1   
16074         0  364348.197352       26.0   среднее             1   
19327         0  389397.167577       26.0    высшее             0   
12681         1  329781.704997       27.0   среднее             1   
...         ...            ...        ...       ...           ...   
229           0  336747.835284       73.0   среднее             1   
4872          0  341528.126150       74.0    высшее             0   
2545          0  372861.103965       74.0   среднее             1   
3445          0  344623.836105       74.0   среднее             1   
19529         0  380150.387046       74.0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
1234   Не женат / не замужем                 4      F   пенсионер     0   
3604         женат / замужем                 0      F   пенсионер     0   
16074       гражданский брак                 1      M   пенсионер     0   
19327        женат / замужем                 0      F   пенсионер     0   
12681        женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
229           вдовец / вдова                 2      F   пенсионер     0   
4872         женат / замужем                 0      F   пенсионер     0   
2545         женат / замужем                 0      F   пенсионер     0   
3445         женат / замужем                 0      M   пенсионер     0   
19529         вдовец / вдова                 2      F   пенсионер     0   

        total_income                           purpose  
1234    89368.600062     получение высшего образования  
3604   110179.690761              покупка своего жилья  
16074   80044.196101                        автомобиль  
19327  214963.301941              покупка недвижимости  
12681   65198.102341             получение образования  
...              ...                               ...  
229    136897.008375  сделка с подержанным автомобилем  
4872   134935.354225              покупка своего жилья  
2545    42927.300898                        автомобили  
3445    54754.745517   операции со своей недвижимостью  
19529   45089.546419           приобретение автомобиля  

[3841 rows x 12 columns]

В данных присутствуют явные недоработки, есть пенсионеры - которым по 22, 24, 26 лет.

**Даже если исключить людей, которые очевидно не входят в число пенсионеров, то у настоящих пенсионеров останутся некорректные значения.** 
* Предположение: Похоже что настоящие значения трудового стажа категории "пенсионеров", каким то образом умножились на значения собственного возраста.

Отсортируем таблицу по возрастанию по стажу.

In [376]:
display(df[df['days_employed'] > max_dob_teor].sort_values(by='days_employed'))

children  days_employed  dob_years education  education_id  \
20327         0  328728.720605       72.0   среднее             1   
9280          2  328734.923996       41.0    высшее             0   
17679         0  328771.341387       56.0   среднее             1   
14699         0  328795.726728       62.0    высшее             0   
7194          1  328827.345667       32.0   среднее             1   
...         ...            ...        ...       ...           ...   
7754          0  401663.850046       61.0   среднее             1   
2145          0  401674.466633       60.0   среднее             1   
7626          1  401675.093434       61.0   среднее             1   
9955          0  401715.811749       69.0    высшее             0   
6922          0  401755.400475       56.0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
20327         вдовец / вдова                 2      F   пенсионер     0   
9280         женат / замужем                 0      M   пенсионер     0   
17679        женат / замужем                 0      F   пенсионер     0   
14699        женат / замужем                 0      F   пенсионер     0   
7194        гражданский брак                 1      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
7754        гражданский брак                 1      F   пенсионер     0   
2145         женат / замужем                 0      M   пенсионер     0   
7626         женат / замужем                 0      F   пенсионер     0   
9955   Не женат / не замужем                 4      F   пенсионер     0   
6922          вдовец / вдова                 2      F   пенсионер     0   

        total_income                                purpose  
20327   96519.339647                покупка жилья для семьи  
9280   126997.497760        операции со своей недвижимостью  
17679   68648.047062  операции с коммерческой недвижимостью  
14699   79940.196752           на покупку своего автомобиля  
7194   122162.965695                        сыграть свадьбу  
...              ...                                    ...  
7754    48286.441362                                свадьба  
2145   325395.724541                             автомобили  
7626   126214.519212                      операции с жильем  
9955    57390.256908                  получение образования  
6922   176278.441171                           ремонт жилью  

[3841 rows x 12 columns]

In [377]:
print('Предполагаемое медианное значение:', int((df.groupby('income_type')['days_employed'].median()[4]) / df.groupby('income_type')['dob_years'].median()[4])) # Медианные значения стажа пенсионеров / медианное значение возраста пенсионеров
print()
print('Предполагаемое(мин.) значение:',int((df[(df.income_type == 'пенсионер') & (df.dob_years >= df.groupby('income_type')['dob_years'].median()[4])]['days_employed'] / df[(df.income_type == 'пенсионер') & (df.dob_years >= df.groupby('income_type')['dob_years'].median()[4])]['dob_years']).min())) # Минимальное значение / возраст клиента
print('Предполагаемое(макс.) значение:',int((df[(df.income_type == 'пенсионер') & (df.dob_years >= df.groupby('income_type')['dob_years'].median()[4])]['days_employed'] / df[(df.income_type == 'пенсионер') & (df.dob_years >= df.groupby('income_type')['dob_years'].median()[4])]['dob_years']).max())) # Максимальное значение / возраст клиента
print()
print('Предполагаемое среднее значение в РФ:', int(247*34.5)) # 247 рабочих дней в году, 34.5 - Средний трудовой стаж в РФ (https://ria.ru/20131119/977986173.html)

Предполагаемое медианное значение: 6010

Предполагаемое(мин.) значение: 4565
Предполагаемое(макс.) значение: 6694

Предполагаемое среднее значение в РФ: 8521


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Чтобы избежать хардкода, старайся вытаскивать нужные значения непосредственно из датафрейма (к примеру, df[df.income_type == 'пенсионер'].days_employed.max())

<div class="alert alert-info">
<b> Комментарий студента:</b> Назначил значения непосредственно из датафрейма

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b>  Абсолютно верно!

Найдем погрешность аномальных данных.

In [378]:
print('Погрешность:',round((df[df['days_employed'] > max_dob_teor]['days_employed'].count() / df['days_employed'].count())*100,2),'%')

Погрешность: 17.95 %


Решение: **Запросить у заказчика новые данные по безработным и пенсионерам. Предупредить о погрешности, если заказчик даст установку продолжать исследование.**


Предположим, что заказчика устраивает такое исследование...
Чистим таблицу, от аномальных значений

In [379]:
df = df[df['days_employed'] < max_dob_teor] # Пересохраняем df согласно условию (41610)
df = df.reset_index(drop=True) # перезаписываем индексы 
df # Итоговая версия таблицы

children  days_employed  dob_years       education  education_id  \
0             1    8437.673028       42.0          высшее             0   
1             1    4024.803754       36.0         среднее             1   
2             0    5623.422610       33.0         Среднее             1   
3             3    4124.747207       32.0         среднее             1   
4             0     926.185831       27.0          высшее             0   
...         ...            ...        ...             ...           ...   
17555         1    2351.431934       37.0  ученая степень             4   
17556         1    4529.316663       43.0         среднее             1   
17557         1    2113.346888       38.0         среднее             1   
17558         3    3112.481705       38.0         среднее             1   
17559         2    1984.507589       40.0         среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      M   компаньон     0   
...                 ...               ...    ...         ...   ...   
17555         в разводе                 3      M   сотрудник     0   
17556  гражданский брак                 1      F   компаньон     0   
17557  гражданский брак                 1      M   сотрудник     1   
17558   женат / замужем                 0      M   сотрудник     1   
17559   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
0      253875.639453                      покупка жилья  
1      112080.014102            приобретение автомобиля  
2      145885.952297                      покупка жилья  
3      267628.550329         дополнительное образование  
4      255763.565419                      покупка жилья  
...              ...                                ...  
17555  115949.039788  покупка коммерческой недвижимости  
17556  224791.862382                  операции с жильем  
17557   89672.561153                       недвижимость  
17558  244093.050500       на покупку своего автомобиля  
17559   82047.418899              на покупку автомобиля  

[17560 rows x 12 columns]

## Шаг 2.3. Изменение типов данных.

Изменяем тип данных. 
Округлим трудовой стаж, возраст и ежемесячный доход до целых чисел.

In [380]:
df['days_employed'] = df['days_employed'].astype('int') # округление струд. стажа
df['dob_years'] = df['dob_years'].astype('int') # округление возраста
df['total_income'] = df['total_income'].astype('int') # округление ежемесячного дохода
df.info() # Проверяем

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17560 entries, 0 to 17559
Data columns (total 12 columns):
children            17560 non-null int64
days_employed       17560 non-null int64
dob_years           17560 non-null int64
education           17560 non-null object
education_id        17560 non-null int64
family_status       17560 non-null object
family_status_id    17560 non-null int64
gender              17560 non-null object
income_type         17560 non-null object
debt                17560 non-null int64
total_income        17560 non-null int64
purpose             17560 non-null object
dtypes: int64(7), object(5)
memory usage: 1.6+ MB


В таблице остался только: int и object.

## Шаг 2.4. Удаление неявных дубликатов.

Изменим регистр.
* `education` — уровень образования клиента, нужно привести к единому регистру.

In [381]:
print(df['education'].unique()) # Вывод уникальных значений

['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'УЧЕНАЯ СТЕПЕНЬ' 'ученая степень']


Привидем к единому регистру методом lower() и проверим результат.

In [382]:
df['education'] = df['education'].str.lower()
print(df['education'].unique())

['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>  Так избавились от неявных дублей, отлично!

## Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим два новых датафрейма со столбцами:
* education_id и education - в первом;
* family_status_id и family_status - во втором.

In [383]:
df_education = df[['education_id','education']] # education_id и education
display(df_education) # Проверка
df_family_status = df[['family_status_id','family_status']] # family_status_id и family_status
display(df_family_status) # Проверка

education_id       education
0                 0          высшее
1                 1         среднее
2                 1         среднее
3                 1         среднее
4                 0          высшее
...             ...             ...
17555             4  ученая степень
17556             1         среднее
17557             1         среднее
17558             1         среднее
17559             1         среднее

[17560 rows x 2 columns]

family_status_id     family_status
0                     0   женат / замужем
1                     0   женат / замужем
2                     0   женат / замужем
3                     0   женат / замужем
4                     1  гражданский брак
...                 ...               ...
17555                 3         в разводе
17556                 1  гражданский брак
17557                 1  гражданский брак
17558                 0   женат / замужем
17559                 0   женат / замужем

[17560 rows x 2 columns]

Удаляем из исходного датафрейма столбы education и family_status, оставив только их идентификаторы.

In [384]:
df = df[['children','days_employed','dob_years','education_id','family_status_id','gender','income_type','debt','total_income','purpose']]
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0            926         27             0                 1      M   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   компаньон     0        255763               покупка жилья

## Шаг 2.6. Категоризация дохода.

Объединим данные по ежемесячному доходу в категории по типу:
* 0–30000 — E;
* 30001–50000 — D;
* 50001–200000 — C;
* 200001–1000000 — B;
* 1000001 и выше — A.

In [385]:
def total_income_category(income): # создаем функции категоризации по доходу
    if income <= 30000:
        return 'E'
    if 30001 <= income <= 50000:
        return 'D'
    if 50001 <= income <= 200000:
        return 'C'
    if 200001 <= income <= 1000000:
        return 'B'
    if income >= 1000001:
        return 'A'
    
df['total_income_category'] = df['total_income'].apply(total_income_category)
df.head()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]


children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0            926         27             0                 1      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   компаньон     0        255763               покупка жилья   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     B

## Шаг 2.7. Категоризация целей кредита.

Объединим данные целей в категории:
* операции с автомобилем;
* операции с недвижимостью;
* проведение свадьбы;
* получение образования;

In [386]:
def purpose_category(list):
    if 'авто' in list:
        return "операции с автомобилем"
    if "образ" in list:
        return "получение образования"
    if "свад" in list:
        return "проведение свадьбы"
    if "недвиж" in list:
        return "операции с недвижимостью"
    if "строи" in list:
        return "операции с недвижимостью"
    if "жил" in list:
        return "операции с недвижимостью"

df['purpose_category'] = df['purpose'].apply(purpose_category)
df = df[['children','days_employed','dob_years','education_id','family_status_id','gender','income_type','debt','total_income_category','purpose_category']]
display(df.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0            926         27             0                 1      M   

  income_type  debt total_income_category          purpose_category  
0   сотрудник     0                     B  операции с недвижимостью  
1   сотрудник     0                     C    операции с автомобилем  
2   сотрудник     0                     C  операции с недвижимостью  
3   сотрудник     0                     B     получение образования  
4   компаньон     0                     B  операции с недвижимостью

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Чтобы избавиться от хардкода, можно заменить данные через if...else:
        
        if "свадьба" in df.purpose:
            df.purpose_category = "свадьба" 
        elif " авто" in df.purpose:
            df.purpose_category = "операции с автомобилем"
        

<div class="alert alert-info">
<b> Комментарий студента:</b> Вау!!! Так намного легче... Странно что после категоризаци, мне не пришел этот метод в голову...

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Получилось компактно и быстро. Отлично!

## Шаг 2.8. Удаление полных дубликатов.

Теперь можно проверить таблицу на наличие дубликатов.

In [475]:
dubl = df[df.duplicated()]
display(dubl)
print('Сумма дубликатов:', dubl['days_employed'].count())
print()
columns = []
for row in dubl:
    #print(dubl[row].head(0)) # вывод названия столбца
    print(dubl[row].value_counts().head(3)) # вывод суммы наиболее популярных дубликатов по столбцам
    dubl_list = dubl[dubl[row] == dubl[row].value_counts().index[0]]
    display(dubl_list.head(5))
    columns.append(row)
    #display(dubl[dubl[columns] == dubl[columns].value_counts().index[0]])

children  days_employed  dob_years  education_id  family_status_id  \
292           0           1311         27             1                 0   
680           0           1311         57             1                 1   
879           0           1360         44             1                 1   
1021          0           1360         54             1                 1   
1819          1           1360         49             0                 0   
...         ...            ...        ...           ...               ...   
17308         0           1360         48             1                 0   
17364         1           1360         30             0                 0   
17432         0           1360         50             1                 0   
17453         0           1311         52             1                 0   
17543         1           1360         42             1                 0   

      gender income_type  debt total_income_category          purpose_category  
292        M   компаньон     0                     C  операции с недвижимостью  
680        F   компаньон     0                     C     получение образования  
879        F   сотрудник     0                     C  операции с недвижимостью  
1021       M   сотрудник     0                     C        проведение свадьбы  
1819       F   сотрудник     0                     C  операции с недвижимостью  
...      ...         ...   ...                   ...                       ...  
17308      F   сотрудник     0                     C    операции с автомобилем  
17364      F   сотрудник     0                     C  операции с недвижимостью  
17432      M   сотрудник     0                     C  операции с недвижимостью  
17453      F   компаньон     0                     C  операции с недвижимостью  
17543      F   сотрудник     0                     C  операции с недвижимостью  

[289 rows x 10 columns]

Сумма дубликатов: 289

0    203
1     67
2     17
Name: children, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
680          0           1311         57             1                 1   
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1926         0           1360         47             1                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
680       F   компаньон     0                     C     получение образования  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1926      M   сотрудник     0                     C  операции с недвижимостью

1360    221
1311     52
2385      9
Name: days_employed, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1819         1           1360         49             0                 0   
1926         0           1360         47             1                 0   
2330         0           1360         41             1                 0   

     gender income_type  debt total_income_category          purpose_category  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1819      F   сотрудник     0                     C  операции с недвижимостью  
1926      M   сотрудник     0                     C  операции с недвижимостью  
2330      F   сотрудник     0                     C  операции с недвижимостью

47    20
42    14
49    14
Name: dob_years, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
1926         0           1360         47             1                 0   
4061         0           1360         47             1                 0   
6879         0           1311         47             1                 0   
7071         0           1360         47             0                 0   
7373         0           1360         47             0                 0   

     gender income_type  debt total_income_category          purpose_category  
1926      M   сотрудник     0                     C  операции с недвижимостью  
4061      M   сотрудник     0                     C  операции с недвижимостью  
6879      F   компаньон     0                     C  операции с недвижимостью  
7071      F   сотрудник     0                     C     получение образования  
7373      F   сотрудник     0                     C  операции с недвижимостью

1    246
0     43
Name: education_id, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
680          0           1311         57             1                 1   
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1926         0           1360         47             1                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
680       F   компаньон     0                     C     получение образования  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1926      M   сотрудник     0                     C  операции с недвижимостью

0    230
1     33
4     23
Name: family_status_id, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
1819         1           1360         49             0                 0   
1926         0           1360         47             1                 0   
2330         0           1360         41             1                 0   
2464         1           1360         26             1                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
1819      F   сотрудник     0                     C  операции с недвижимостью  
1926      M   сотрудник     0                     C  операции с недвижимостью  
2330      F   сотрудник     0                     C  операции с недвижимостью  
2464      F   сотрудник     0                     C  операции с недвижимостью

F    202
M     87
Name: gender, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
680          0           1311         57             1                 1   
879          0           1360         44             1                 1   
1819         1           1360         49             0                 0   
2330         0           1360         41             1                 0   
2464         1           1360         26             1                 0   

     gender income_type  debt total_income_category          purpose_category  
680       F   компаньон     0                     C     получение образования  
879       F   сотрудник     0                     C  операции с недвижимостью  
1819      F   сотрудник     0                     C  операции с недвижимостью  
2330      F   сотрудник     0                     C  операции с недвижимостью  
2464      F   сотрудник     0                     C  операции с недвижимостью

сотрудник      226
компаньон       54
госслужащий      9
Name: income_type, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1819         1           1360         49             0                 0   
1926         0           1360         47             1                 0   
2330         0           1360         41             1                 0   

     gender income_type  debt total_income_category          purpose_category  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1819      F   сотрудник     0                     C  операции с недвижимостью  
1926      M   сотрудник     0                     C  операции с недвижимостью  
2330      F   сотрудник     0                     C  операции с недвижимостью

0    287
1      2
Name: debt, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
680          0           1311         57             1                 1   
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1819         1           1360         49             0                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
680       F   компаньон     0                     C     получение образования  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1819      F   сотрудник     0                     C  операции с недвижимостью

C    288
B      1
Name: total_income_category, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
680          0           1311         57             1                 1   
879          0           1360         44             1                 1   
1021         0           1360         54             1                 1   
1819         1           1360         49             0                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
680       F   компаньон     0                     C     получение образования  
879       F   сотрудник     0                     C  операции с недвижимостью  
1021      M   сотрудник     0                     C        проведение свадьбы  
1819      F   сотрудник     0                     C  операции с недвижимостью

операции с недвижимостью    193
получение образования        40
операции с автомобилем       34
Name: purpose_category, dtype: int64


children  days_employed  dob_years  education_id  family_status_id  \
292          0           1311         27             1                 0   
879          0           1360         44             1                 1   
1819         1           1360         49             0                 0   
1926         0           1360         47             1                 0   
2330         0           1360         41             1                 0   

     gender income_type  debt total_income_category          purpose_category  
292       M   компаньон     0                     C  операции с недвижимостью  
879       F   сотрудник     0                     C  операции с недвижимостью  
1819      F   сотрудник     0                     C  операции с недвижимостью  
1926      M   сотрудник     0                     C  операции с недвижимостью  
2330      F   сотрудник     0                     C  операции с недвижимостью

In [474]:
dubl[
    (dubl[columns[0]]==dubl[columns[0]].value_counts().index[0])&
    (dubl[columns[1]]==dubl[columns[1]].value_counts().index[0])&
    (dubl[columns[2]]==dubl[columns[2]].value_counts().index[0])&
    (dubl[columns[3]]==dubl[columns[3]].value_counts().index[0])&
    (dubl[columns[4]]==dubl[columns[4]].value_counts().index[0])&
    (dubl[columns[5]]==dubl[columns[5]].value_counts().index[0])&
    (dubl[columns[6]]==dubl[columns[6]].value_counts().index[0])&
    (dubl[columns[7]]==dubl[columns[7]].value_counts().index[0])&
    (dubl[columns[8]]==dubl[columns[8]].value_counts().index[0])&
    (dubl[columns[9]]==dubl[columns[9]].value_counts().index[0])
] # Пример дубликатов по наибольшим параметрам


children  days_employed  dob_years  education_id  family_status_id  \
7397          0           1360         47             1                 0   
8064          0           1360         47             1                 0   
9624          0           1360         47             1                 0   
9660          0           1360         47             1                 0   
11787         0           1360         47             1                 0   
13197         0           1360         47             1                 0   
14480         0           1360         47             1                 0   
17241         0           1360         47             1                 0   

      gender income_type  debt total_income_category          purpose_category  
7397       F   сотрудник     0                     C  операции с недвижимостью  
8064       F   сотрудник     0                     C  операции с недвижимостью  
9624       F   сотрудник     0                     C  операции с недвижимостью  
9660       F   сотрудник     0                     C  операции с недвижимостью  
11787      F   сотрудник     0                     C  операции с недвижимостью  
13197      F   сотрудник     0                     C  операции с недвижимостью  
14480      F   сотрудник     0                     C  операции с недвижимостью  
17241      F   сотрудник     0                     C  операции с недвижимостью

289 дубликатов. Удаляем их.

In [479]:
df = df.drop_duplicates()
#df
df.duplicated().sum()

0

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Явные дубли - это полные совпадения строк во всем датафрейме. Давай найдем их!

<div class="alert alert-info">
<b> Комментарий студента:</b> Исправил!

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Избавилсиь от дублей!

**Итоговый датафрейм**

In [480]:
df = df.reset_index(drop=True)
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0            926         27             0                 1   
...         ...            ...        ...           ...               ...   
17266         1           2351         37             4                 3   
17267         1           4529         43             1                 1   
17268         1           2113         38             1                 1   
17269         3           3112         38             1                 0   
17270         2           1984         40             1                 0   

      gender income_type  debt total_income_category          purpose_category  
0          F   сотрудник     0                     B  операции с недвижимостью  
1          F   сотрудник     0                     C    операции с автомобилем  
2          M   сотрудник     0                     C  операции с недвижимостью  
3          M   сотрудник     0                     B     получение образования  
4          M   компаньон     0                     B  операции с недвижимостью  
...      ...         ...   ...                   ...                       ...  
17266      M   сотрудник     0                     C  операции с недвижимостью  
17267      F   компаньон     0                     B  операции с недвижимостью  
17268      M   сотрудник     1                     C  операции с недвижимостью  
17269      M   сотрудник     1                     B    операции с автомобилем  
17270      F   сотрудник     0                     C    операции с автомобилем  

[17271 rows x 10 columns]

## Шаг 3. Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Что бы ответить на этот вопрос, узнаем у скольких клиентов с детьми были просрочки по кредиту и разделим на количество клиентов с детьми, которые платили вовремя.

In [494]:
print('Общая зависимость:',round((df[df['debt'] == 1]['children'].count() / df[df['debt'] == 0]['children'].count())*100,2),'%')

Общая зависимость: 9.6 %


In [501]:
print('Зависимость без детей:',round((df[(df['debt'] == 1) & (df['children'] == 0)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 0)]['children'].count())*100,2),'%')
print('Зависимость при одним ребенком:',round((df[(df['debt'] == 1) & (df['children'] == 1)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 1)]['children'].count())*100,2),'%')
print('Зависимость при вдух детях:',round((df[(df['debt'] == 1) & (df['children'] == 2)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 2)]['children'].count())*100,2),'%')
print('Зависимость при трех детях:',round((df[(df['debt'] == 1) & (df['children'] == 3)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 3)]['children'].count())*100,2),'%')
print('Зависимость при четырех детях:',round((df[(df['debt'] == 1) & (df['children'] == 4)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 4)]['children'].count())*100,2),'%')
print('Зависимость при пяти детях:',round((df[(df['debt'] == 1) & (df['children'] == 5)]['children'].count() / df[(df['debt'] == 0) & (df['children'] == 5)]['children'].count())*100,2),'%')

Зависимость без детей: 9.04 %
Зависимость при одним ребенком: 10.6 %
Зависимость при вдух детях: 10.45 %
Зависимость при трех детях: 8.75 %
Зависимость при четырех детях: 11.43 %
Зависимость при пяти детях: 0.0 %


Интересно что при пяти детях у семей, просрочка по кредиту 0%

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Здесь я бы советовала создать сводную таблицу и сгруппировать данные по среднему. Так мы сразу увилим долю

<div class="alert alert-info">
<b> Комментарий студента:</b> Цифры другие. Что сделал неправильно, подскажите пожалуйста))

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b>  Как можно сформировать таблицу:
        
            df.pivot_table(index="cchildren",columns='debt', values="education", aggfunc="mean")
            
Таким образом, мы сразу получим процентное соотношение оптимальным способом.

При твоем расчете доли нужно сделать следующее:
    
        len(df[(df['debt'] == 1) & (df['children'] == 3)])/len(df['children'] == 3)
        
То есть долю людей с определенным кол-вом детей и просрочкой делим на общую долю людей с определенным кол-вом детей
            

In [503]:
df.pivot_table(index='children', values ='debt' , aggfunc = 'mean')

debt
children          
0         0.082925
1         0.095866
2         0.094648
3         0.080495
4         0.102564
5         0.000000

In [491]:
print('Средний процент просрочки с детьми',round((10.38 + 10.37 + 8.72 + 11.11)/5,2), '%')

Средний процент просрочки с детьми 8.12 %


### Вывод 1: В среднем, ребенок увеличивает возможность просрочки платежа по кредиту.

### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Что бы ответить на этот вопрос, узнаем скольких клиентов были просрочки и разделим на количество клиентов, которые платили вовремя, учитывая семейное положение клиента.

In [505]:
print('Зависимость при браке:',round((df[(df['debt'] == 1) & (df['family_status_id'] == 0)]['family_status_id'].count() / df[(df['debt'] == 0) & (df['family_status_id'] == 0)]['family_status_id'].count())*100,2),'%')
print('Зависимость при гражданском браке:',round((df[(df['debt'] == 1) & (df['family_status_id'] == 1)]['family_status_id'].count() / df[(df['debt'] == 0) & (df['family_status_id'] == 1)]['family_status_id'].count())*100,2),'%')
print('Зависимость при вдовец/вдова:',round((df[(df['debt'] == 1) & (df['family_status_id'] == 2)]['family_status_id'].count() / df[(df['debt'] == 0) & (df['family_status_id'] == 2)]['family_status_id'].count())*100,2),'%')
print('Зависимость при разводе:',round((df[(df['debt'] == 1) & (df['family_status_id'] == 3)]['family_status_id'].count() / df[(df['debt'] == 0) & (df['family_status_id'] == 3)]['family_status_id'].count())*100,2),'%')
print('Зависимость при холост:',round((df[(df['debt'] == 1) & (df['family_status_id'] == 4)]['family_status_id'].count() / df[(df['debt'] == 0) & (df['family_status_id'] == 4)]['family_status_id'].count())*100,2),'%')

Зависимость при браке: 8.82 %
Зависимость при гражданском браке: 11.19 %
Зависимость при вдовец/вдова: 6.63 %
Зависимость при разводе: 7.91 %
Зависимость при холост: 11.85 %


### Вывод 2: Холостые и люди состоящие в гражданском браке имеют больше просрочек.

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Что бы ответить на этот вопрос, узнаем у скольких клиентов были просрочки и разделим на количество клиентов, которые платили вовремя, учитывая их уровень дохода по категориям.

In [489]:
print('Категория E:',round((df[(df['debt'] == 1) & (df['total_income_category'] == 'E')]['total_income_category'].count() / df[(df['debt'] == 0) & (df['total_income_category'] == 'E')]['total_income_category'].count())*100,2),'%')
print('Категория D:',round((df[(df['debt'] == 1) & (df['total_income_category'] == 'D')]['total_income_category'].count() / df[(df['debt'] == 0) & (df['total_income_category'] == 'D')]['total_income_category'].count())*100,2),'%')
print('Категория C:',round((df[(df['debt'] == 1) & (df['total_income_category'] == 'C')]['total_income_category'].count() / df[(df['debt'] == 0) & (df['total_income_category'] == 'C')]['total_income_category'].count())*100,2),'%')
print('Категория B:',round((df[(df['debt'] == 1) & (df['total_income_category'] == 'B')]['total_income_category'].count() / df[(df['debt'] == 0) & (df['total_income_category'] == 'B')]['total_income_category'].count())*100,2),'%')
print('Категория A:',round((df[(df['debt'] == 1) & (df['total_income_category'] == 'A')]['total_income_category'].count() / df[(df['debt'] == 0) & (df['total_income_category'] == 'A')]['total_income_category'].count())*100,2),'%')

Категория E: 0.0 %
Категория D: 8.19 %
Категория C: 10.28 %
Категория B: 7.81 %
Категория A: 8.7 %


Интересно что клиенты категории Е выплачивают все как положенно.

### Вывод 3: У клиентов категории С (ежемесячный доход которых 50001–200000) - чаще встречаются просрочки.

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

Что бы ответить на этот вопрос, узнаем у скольких клиентов были просрочки и разделим на количество клиентов, которые платили вовремя, учитывая их цели кредита.

In [490]:
print('операции с автомобилем:',round((df[(df['debt'] == 1) & (df['purpose_category'] == 'операции с автомобилем')]['purpose_category'].count() / df[(df['debt'] == 0) & (df['purpose_category'] == 'операции с автомобилем')]['purpose_category'].count())*100,2),'%')
print('операции с недвижимостью:',round((df[(df['debt'] == 1) & (df['purpose_category'] == 'операции с недвижимостью')]['purpose_category'].count() / df[(df['debt'] == 0) & (df['purpose_category'] == 'операции с недвижимостью')]['purpose_category'].count())*100,2),'%')
print('проведение свадьбы:',round((df[(df['debt'] == 1) & (df['purpose_category'] == 'проведение свадьбы')]['purpose_category'].count() / df[(df['debt'] == 0) & (df['purpose_category'] == 'проведение свадьбы')]['purpose_category'].count())*100,2),'%')
print('получение образования:',round((df[(df['debt'] == 1) & (df['purpose_category'] == 'получение образования')]['purpose_category'].count() / df[(df['debt'] == 0) & (df['purpose_category'] == 'получение образования')]['purpose_category'].count())*100,2),'%')

операции с автомобилем: 11.18 %
операции с недвижимостью: 8.57 %
проведение свадьбы: 9.16 %
получение образования: 11.03 %


### Вывод 4: У клиентов, которые брали кредит на операции с автомобилем и на образование - чаще встречаются просрочки.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Все так

## Общий вывод:

При работе с данными были заполнены пропущенные значения, удалены дубликаты и аномальные данные, выделены категории ежемесячного дохода и целей кредита. Дополнительный анализ может помочь более точно предсказать вероятность вылпаты кредита в срок.
Из-за отсутствия 17.92% данных. В дальнейшем результаты выводов могут измениться.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Спасибо за работу! Проект получился очень хорошим. Давай сделаем его сильнее:
        
- удалим дявные убли из всей таблицы
- оптимизируем код и избавимся от хардкода

Оставила направляющие комментарии. Возвращаю тебе работу, чтобы ты смог задать вопросы. Буду ждать твою работу на финальное ревью!
</div>

<div class="alert alert-info">
<b> Комментарий студента:</b> Исправил некоторые моменты, воспользовался вашими советами!

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Спасибо за проработку комментариев! Теперь все отлично. Оставила еще направляющих комментариев. Если у тебя возникнут вопросы, смело задавай мне их через куратора. Удачи на следующих этапах!